addons/sale_mrp/models/sale.py

```diff
@@ -153,10 +153,16 @@ def _get_bom_component_qty(self, bom):
     def _get_qty_procurement(self, previous_product_uom_qty=False):
         self.ensure_one()
         # Specific case when we change the qty on a SO for a kit product.
-        # We don't try to be too smart and keep a simple approach: we compare the quantity before
-        # and after update, and return the difference. We don't take into account what was already
-        # sent, or any other exceptional case.
+        # We don't try to be too smart and keep a simple approach: we use the quantity of entire
+        # kits that are currently in delivery
         bom = self.env['mrp.bom']._bom_find(self.product_id, bom_type='phantom')[self.product_id]
         if bom:
-            return previous_product_uom_qty and previous_product_uom_qty.get(self.id, 0.0) or self.qty_delivered
+            moves = self.move_ids.filtered(lambda r: r.state != 'cancel' and not r.scrapped)
+            filters = {
+                'incoming_moves': lambda m: m.location_dest_id.usage == 'customer' and (not m.origin_returned_move_id or (m.origin_returned_move_id and m.to_refund)),
+                'outgoing_moves': lambda m: m.location_dest_id.usage != 'customer' and m.to_refund
+            }
+            order_qty = self.product_uom._compute_quantity(self.product_uom_qty, bom.product_uom_id)
+            qty = moves._compute_kit_quantities(self.product_id, order_qty, bom, filters)
```

Comment:
Something is being miscalculated here. Try doing the same steps but put the SOL in dozens but leave everything else as units and you will end up with with a return with weirdly large numbers. :thinking:

Reply:
damn it, forgot to revert the qty conversion
good catch, thx
I have updated the fix and included this in the use case of the test